tensorflow/contrib/periodic_resample/kernels/periodic_resample_op.h

```diff
@@ -163,29 +285,87 @@ template <class InputDataT,
   for (int i = 0; i < rank; ++i) {
     output_shape.AddDim(target_dimensions[i]);
   }
-  const auto new_size =
-      new_sliced_size * target_dimensions[adjustable_dimension];
 
   // Create an output tensor and attach it to the current context
   tensorflow::Tensor* output_tensor = nullptr;
   OP_REQUIRES_OK(context,
                  context->allocate_output(0, output_shape, &output_tensor));
   auto output = output_tensor->flat<InputDataT>();
 
-  // memory is allocated for these variables outside the inner loop for
-  // efficiency (although, I could create a separate class scope for
-  // this purpose instead)
-  tensorflow::int64 result = 0;
-  std::vector<tensorflow::int64> output_indices(target_dimensions.size());
+  // Fill output tensor with periodically resampled input tensor values
+  InputIndexer input_indexer(target_dimensions, input_tensor.shape(),
+                             adjustable_dimension);
+
+  auto worker_threads = *(context->device()->tensorflow_cpu_worker_threads());
+  auto fill_output_tensor = [&input_indexer, &output, &input](
+      tensorflow::int64 start, tensorflow::int64 limit) {
+    InputIndexer local_indexer(input_indexer);
+    local_indexer.MoveToOutputIndex(start);
+    for (tensorflow::int64 output_index = start; output_index < limit;
+         ++output_index) {
+      output(output_index) = input(local_indexer.linear_input_index());
+      local_indexer.IncrementOutputIndex();
+    }
+  };
+  ::tensorflow::Shard(worker_threads.num_threads, worker_threads.workers,
+                      new_size, costPerFillIndex, fill_output_tensor);
+}
+
+template <class InputDataT>
+void fill_grad_tensor(tensorflow::OpKernelContext* context,
```

Comment:
It seems like this function could share most of the code with fill_prediodic_tensor. Can you try to consolidate?

Reply:
Done